src/animation.rs

```diff
@@ -0,0 +1,503 @@
+    }
+
+    /// Disables the animation action.
+    pub fn disable(&mut self) -> &mut Self {
+        self.send(Operation::Disable)
+    }
+
+    /// Pauses the animation action.
+    pub fn pause(&mut self) -> &mut Self {
+        self.send(Operation::Pause)
+    }
+
+    /// Plays the animation action.
+    pub fn play(&mut self) -> &mut Self {
+        self.send(Operation::Play)
+    }
+
+    /// Sets the animation loop mode.
+    pub fn set_loop_mode(
+        &mut self,
+        loop_mode: LoopMode,
+    ) -> &mut Self {
+        self.send(Operation::SetLoopMode(loop_mode))
+    }
+}
+
+impl Mixer {
+    fn process_messages(&mut self) {
+        while let Ok((weak_ptr, operation)) = self.rx.try_recv() {
+            let mut action = match weak_ptr.upgrade() {
+                Ok(ptr) => &mut self.actions[&ptr],
+                Err(_) => continue,
+            };
+            match operation {
+                Operation::Enable => action.enabled = true,
+                Operation::Disable => action.enabled = false,
+                Operation::Pause => action.paused = true,
+                Operation::Play => {
+                    action.paused = false;
+                    action.enabled = true;
+                }
+                Operation::SetLoopMode(loop_mode) => action.loop_mode = loop_mode,
+            }
+        }
+    }
+
+    fn update_actions(
+        &mut self,
+        delta_time: f32,
+    ) {
+        for mut action in self.actions.iter_mut() {
+            action.update(delta_time);
+        }
+    }
+
+    /// Creates a new animation mixer.
+    pub fn new() -> Self {
+        let actions = froggy::Storage::new();
+        let (tx, rx) = mpsc::channel();
+        Mixer { actions, rx, tx }
+    }
+
+    /// Spawns a new animation [`Action`] to be updated by this mixer.
+    ///
+    /// [`Action`]: struct.Action.html
+    pub fn action(
+        &mut self,
+        clip: Clip,
+    ) -> Action {
+        let action_data = ActionData::new(clip);
+        let pointer = self.actions.create(action_data);
+        let tx = self.tx.clone();
+        Action { tx, pointer }
+    }
+
+    /// Updates the actions owned by the mixer.
+    pub fn update(
+        &mut self,
+        delta_time: f32,
+    ) {
+        self.process_messages();
+        self.update_actions(delta_time);
+    }
+}
+
+impl ActionData {
+    fn new(clip: Clip) -> Self {
+        ActionData {
+            clip: clip,
+            enabled: true,
+            loop_mode: LoopMode::Repeat(Repetitions::Infinite),
+            paused: false,
+            local_time: 0.0,
+            local_time_scale: 1.0,
+        }
+    }
+
+    /// Updates a single animation action.
+    fn update(
+        &mut self,
+        delta_time: f32,
+    ) {
+        if self.paused || !self.enabled {
+            return;
+        }
+
+        self.local_time += delta_time * self.local_time_scale;
+        let mut finish_count = 0;
+        for mut track in &mut self.clip.tracks {
```

Comment:
why do we mutate the clip tracks?

Reply:
Looks like an oversight. I'll change that.

Reply:
It's not an oversight. It's because `Track` holds the `Object` it updates. I'm not sure how else to implement this.

Reply:
We need a way to separate immutable stuff from mutable...